File: 2025-10-October-GFG-POTD/October 22 - Nearly sorted/Solution.py
import heapq

class Solution:
    def nearlySorted(self, arr, k):
        # Create a min heap
        minHeap = []
        n = len(arr)
        index = 0

        # Step 1: Add first k+1 elements to heap
        for i in range(min(k + 1, n)):
            heapq.heappush(minHeap, arr[i])

        # Step 2: Process remaining elements
        for i in range(k + 1, n):
            arr[index] = heapq.heappop(minHeap)
            index += 1
            heapq.heappush(minHeap, arr[i])

        # Step 3: Extract remaining elements
        while minHeap:
            arr[index] = heapq.heappop(minHeap)
            index += 1

File: 2025-10-October-GFG-POTD/October 22 - Nearly sorted/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void nearlySorted(vector<int>& arr, int k) {
        // Create a min heap
        priority_queue<int, vector<int>, greater<int>> minHeap;

        int n = arr.size();
        int index = 0;

        // Step 1: Add first k+1 elements to heap
        for (int i = 0; i <= k && i < n; i++) {
            minHeap.push(arr[i]);
        }

        // Step 2: Process remaining elements
        for (int i = k + 1; i < n; i++) {
            arr[index++] = minHeap.top();
            minHeap.pop();
            minHeap.push(arr[i]);
        }

        // Step 3: Extract remaining elements
        while (!minHeap.empty()) {
            arr[index++] = minHeap.top();
            minHeap.pop();
        }
    }
};
